### DISCUSSION TITLE
Switch from local uploads to S3

### DISCUSSION THREAD

**Discussion by Lucas-Gruber** [directus/directus#11083] [Archive]
I've always used the local file storage until now and would now switch to an S3 bucket, how can I do that without having to relink every image (in every post or link)?

**Comment by azrikahar** [comment]
Provided you upload the files in the `uploads` folder as-is (with their UUID intact) into the S3 bucket, it should work. 👍

**Reply by joselcvarela** [reply]
Marking this as answer, but let us know if that worked @Thomas-Blakeney

**Reply by rijkvanzanten** [reply]
Also make sure to keep the name of the storage adapter `local`, or update all the records in `directus_files` to use the new name of the storage adapter 👍🏻

**Reply by Lucas-Gruber** [reply]
Where can i create client secret and key for the .env File?
And is it required to set storage locations local and s3 or only s3?

**Reply by joselcvarela** [reply]
Check if this can help
https://docs.directus.io/configuration/config-options/#file-storage
and let us know.

**Reply by Lucas-Gruber** [reply]
Leider funktioniert es nicht ohne lokalen Speicher, nur wenn ich beides konfiguriere, s3 UND lokal, funktioniert es, nur neue Dateien werden auf s3 hochgeladen, aber alle Dateien, die manuell in den Speicher-Bucket hochgeladen werden, funktionieren nicht.

>Translated
>Unfortunately it doesn't work without local storage, only when I configure both s3 AND local does it work, only new files are uploaded to s3, but any files uploaded manually to the storage bucket don't work.

![image](https://user-images.githubusercontent.com/72545889/149809453-2c4bde0a-b463-4ce1-b05f-844cb4a0c9e9.png)

Can you maybe help me?

**Reply by joselcvarela** [reply]
Try to keep with English so everyone can understand 👍
Files uploaded manually to S3 will not work because we need to keep a record of all files, so you must upload everything via Directus.

Regarding working exclusively with S3, you may want to create an issue for that, since it should work with at least a single storage provider.

**Reply by Lucas-Gruber** [reply]
Okay, but when i uploaded the Files from the "Uploads" folder the Files has the same names.. but is nothing way to transfer at this version?

**Reply by joselcvarela** [reply]
What do you mean with `transfer`?
As long as you upload the files from your `uploads` folder to S3 without any changes it should work.

**Reply by joselcvarela** [reply]
Take a look into this answer, it may help https://github.com/directus/directus/discussions/10860#discussioncomment-1915858

**Reply by Lucas-Gruber** [reply]
> What do you mean with `transfer`? As long as you upload the files from your `uploads` folder to S3 without any changes it should work.

Yes, but I've been using local storage for a long time, but would like to switch to S3 afterwards

**Reply by rijkvanzanten** [reply]
> Yes, but I've been using local storage for a long time, but would like to switch to S3 afterwards

As long as you follow the following, you should be OK:

1) Copy all uploaded files from your local folder to the S3 bucket
2) Change the configuration from:

```
STORAGE_LOCATIONS="local"
STORAGE_LOCAL_DRIVER="local"
STORAGE_LOCAL_ROOT="./uploads"
```

to

```
STORAGE_LOCATIONS="local"
STORAGE_LOCAL_DRIVER="s3"
STORAGE_LOCAL_KEY / SECRET / BUCKET / etc
```

If you want to rename your "local" storage location to something else, like "cloud", make sure to update every record in `directus_files` to have `storage = "cloud"` instead of `storage = "local"`.

**Reply by Lucas-Gruber** [reply]
I've already done that but unfortunately it doesn't find the files, only the ones that I upload afterwards.

Although all (including the freshly uploaded) images are in the bucket

**Comment by machariamuguku** [comment]
Commenting here for anyone who's looking to migrate from local storage to s3. Here's what worked for me:

1. Copy the files in your local uploads
	```bash
	# E.g with the AWS CLI 
	
	# Install AWS CLI
	sudo apt install awscli (or sudo snap install aws-cli --classic)
	
	# Configure bucket access user
	# Supply `AWS Access Key ID` and `AWS Secret Access Key` for the s3 bucket user when prompted
	aws configure
	
	# Copy files recursively 
	aws s3 cp ./uploads s3://<bucket-name>/<uploads-directory>/ --recursive
	```

2. Change the local configs:
	```env
	# from
	STORAGE_LOCATIONS="local"
	STORAGE_LOCAL_DRIVER="local"
	STORAGE_LOCAL_ROOT="./uploads"
	```

	```env
	# to
	STORAGE_LOCATIONS="s3"
	STORAGE_S3_DRIVER="s3"
	# Other STORAGE_S3_<CONFIGS> listed here: https://directus.io/docs/configuration/files#s3-s3
	```

3. Update the `directus_files` metadata table (system collection) to change the values of the `storage` column for all file entries from `local` to `s3`:

	```sql
	UPDATE directus_files
	SET storage = 's3'
	WHERE storage = 'local';
	```

**Comment by gholleyd** [comment]
More hints:

Download your old ./uploads dir. Upload it to your new S3 bucket. Using something like rclone makes it easy. https://developers.cloudflare.com/r2/objects/upload-objects/#upload-objects-via-rclone

This config works:

```
STORAGE_LOCATIONS: "s3,local"

STORAGE_LOCAL_DRIVER: "s3"
STORAGE_LOCAL_KEY: "[ACCESS_KEY_ID]"
STORAGE_LOCAL_SECRET: "[SECRET_ACCESS_KEY]"
STORAGE_LOCAL_BUCKET: "[BUCKET]"
STORAGE_LOCAL_REGION: "[REGION]"
STORAGE_LOCAL_ENDPOINT: "[ENDPOINT]"
STORAGE_LOCAL_ROOT: "" # Important! Directus uses ./uploads by default, so make sure to over write it to the correct root folder on S3.

STORAGE_S3_DRIVER: "s3"
STORAGE_S3_KEY: "[ACCESS_KEY_ID]"
STORAGE_S3_SECRET: "[SECRET_ACCESS_KEY]"
STORAGE_S3_BUCKET: "[BUCKET]"
STORAGE_S3_REGION: "[REGION]"
STORAGE_S3_ENDPOINT: "[ENDPOINT]"
STORAGE_S3_ROOT: ""
```

Alternatively, you can also leave all of the old local files right where they are and just send new files to S3. Directus understands.

```
STORAGE_LOCATIONS: "s3,local"

STORAGE_S3_DRIVER: "s3"
STORAGE_S3_KEY: "[ACCESS_KEY_ID]"
STORAGE_S3_SECRET: "[SECRET_ACCESS_KEY]"
STORAGE_S3_BUCKET: "[BUCKET]"
STORAGE_S3_REGION: "[REGION]"
STORAGE_S3_ENDPOINT: "[ENDPOINT]"
STORAGE_S3_ROOT: ""
```
